### PR TITLE
Kernel: Convert TCP pseudo-headers through a union

### LINKED ISSUE #14004: Kernel: TCPSocket miscalculates checksums if `-fsanitize=undefined` is disabled
The `Kernel::TCPSocket::compute_tcp_checksum()` function apparently returns wrong results if `-fsanitize=undefined` is disabled. This is noticeable when actually trying to establish a TCP connection, which will result in packages being dropped because the checksum doesn't match the contents.

Specifying `-fno-strict-aliasing` for said function resolves or works around the issue, as TCP connections start to work again.

Only GCC seems to be affected, Clang-based builds continue to successfully establish TCP connections even with UBSAN disabled and without the `-fno-strict-aliasing` workaround.

### PR DESCRIPTION
**Pull request by timschumi**
This keeps us from tripping strict aliasing, which previously made TCP connections inoperable when building without `-fsanitize=undefined` or `-fno-strict-aliasing`.

Fixes #14004.

### CODE REVIEW

**Comment by IdanHo** [comment]
fsanitize=undefined, not address, no?

**Comment by timschumi** [comment]
> fsanitize=undefined, not address, no?

Whoops. Of course, yes.

File: Kernel/Net/TCPSocket.cpp [hunk 1]
```diff
@@ -360,18 +360,22 @@ bool TCPSocket::should_delay_next_ack() const
...
+        u16 raw[6];
     };
+    static_assert(sizeof(PseudoHeader) == 12);
 
-    PseudoHeader pseudo_header { source, destination, 0, (u8)IPv4Protocol::TCP, packet.header_size() + payload_size };
+    PseudoHeader pseudo_header { .header = { source, destination, 0, (u8)IPv4Protocol::TCP, packet.header_size() + payload_size } };
 
     u32 checksum = 0;
-    auto raw_pseudo_header = bit_cast<u16*>(&pseudo_header);
+    auto raw_pseudo_header = pseudo_header.raw;
```
Comment (IdanHo) [inline Kernel/Net/TCPSocket.cpp:378]: auto*

**Review by IdanHo — COMMENTED** [review]
Nice find. Weird that the compiler doesn't warn about that one.

File: Kernel/Net/TCPSocket.cpp [hunk 2]
```diff
@@ -360,18 +360,22 @@ bool TCPSocket::should_delay_next_ack() const
...
+            u8 zero;
+            u8 protocol;
+            NetworkOrdered<u16> payload_size;
+        } header;
+        u16 raw[6];
     };
+    static_assert(sizeof(PseudoHeader) == 12);
 
-    PseudoHeader pseudo_header { source, destination, 0, (u8)IPv4Protocol::TCP, packet.header_size() + payload_size };
+    PseudoHeader pseudo_header { .header = { source, destination, 0, (u8)IPv4Protocol::TCP, packet.header_size() + payload_size } };
```
Comment (ADKaster) [inline Kernel/Net/TCPSocket.cpp:375]: Pre-existing, but the addition for the size parameter could trivially overflow a u16 -- especially with the integer promotion happening as a result of operator+. Do we verify that won't happen anywhere further up the call stack?

File: Kernel/Net/TCPSocket.cpp (hunk 1 above)
Comment (timschumi) [reply Kernel/Net/TCPSocket.cpp:378]: Should we make the others `auto*` as well then? Or just the one I touched?

File: Kernel/Net/TCPSocket.cpp (hunk 2 above)
Comment (timschumi) [reply Kernel/Net/TCPSocket.cpp:375]: Kind of, maybe.

The only place where we calculate the checksum is when we actually send packages, and the only packages that are sent with a payload are in `protocol_send`. That specific function always limits the payload size to the MTU.

It's not an outright verification, but that MTU limit should place the size of an individual package way below the critical limit for anything that we support as of now.

File: Kernel/Net/TCPSocket.cpp (hunk 1 above)
Comment (IdanHo) [reply Kernel/Net/TCPSocket.cpp:378]: As long as it's not too many, doing all of them should be fine.

File: Kernel/Net/TCPSocket.cpp (hunk 2 above)
Comment (awesomekling) [reply Kernel/Net/TCPSocket.cpp:375]: Seems like something we could `VERIFY()` here?

**Pull request merged**